### ISSUE TITLE
Handling more than 10 components

### ISSUE THREAD

**Issue by Nazariglez** [leudz/shipyard#212]
Hi! I’m running into the 10 component limit with BulkAddEntity and system queries in Shipyard (comes from the macro). Is there a way to handle entities or queries with more than 10 components? Any guidance or best practices would be greatly appreciated. 

Thanks!

PD: just to show what I mean:
--
```
macro_rules! bulk_insert {
    ($(($type: ident, $sparse_set: ident, $index: tt))*;($type1: ident, $sparse_set1: ident, $index1: tt) $(($queue_type: ident, $queue_sparse_set: ident, $queue_index: tt))*) => {
        impl_bulk_insert![$(($type, $sparse_set, $index))*];
        bulk_insert![$(($type, $sparse_set, $index))* ($type1, $sparse_set1, $index1); $(($queue_type, $queue_sparse_set, $queue_index))*];
    };
    ($(($type: ident, $sparse_set: ident, $index: tt))*;) => {
        impl_bulk_insert![$(($type, $sparse_set, $index))*];
    }
}

bulk_insert![(A, sparse_set0, 0) (B, sparse_set1, 1); (C, sparse_set2, 2) (D, sparse_set3, 3) (E, sparse_set4, 4) (F, sparse_set5, 5) (G, sparse_set6, 6) (H, sparse_set7, 7) (I, sparse_set8, 8) (J, sparse_set9, 9)];
```

This is not valid unless I remove a component:
```
fn create_boids(world: &mut World, team: Team, n: u64, center: Vec2, radius: f32) {
    let mut next_id = BOID_ID
        .fetch_update(Ordering::SeqCst, Ordering::SeqCst, |v| Some(v + n))
        .unwrap_ctx("Updating BOID_ID");

    world.bulk_add_entity((0..n).map(|i| {
        let id = next_id + i;
        let pos = (center + radial_random_pos(radius)).round();

        let dir = Vec2::new(
            random::gen::<f32>() * 2.0 - 1.0,
            random::gen::<f32>() * 2.0 - 1.0,
        )
        .normalize_or_zero();

        let boid = Boid {
            id,
            team,
            ..Default::default()
        };

        (
            boid,
            Pos(pos),
            BoidDir(dir),
            BoidEnergy(random::range(0.0..boid.max_energy)),
            NearBoids::default(),
            PrayForSpirit::default(),
            AvoidanceForce(Vec2::ZERO),
            CohesionForce(Vec2::ZERO),
            AlignmentForce(Vec2::ZERO),
            SeparationForce(Vec2::ZERO),
            Drawable::Soul { anim_delta: 0.0 },
        )
    }));
}
```

**Comment by leudz** [comment]
Hi! For systems you have a few options.

You can nest tuples. Instead of:
```rust
fn sys(
    a: View<A>,
    b: View<B>,
    c: View<C>,
    d: View<D>,
    e: View<E>,
    f: View<F>,
    g: View<G>,
    h: View<H>,
    i: View<I>,
    j: View<J>,
    k: View<K>,
) {}
```

You'd have something like this:
```rust
fn sys(
    (a, b, c, d, e, f, g, h, i, j): (
        View<A>,
        View<B>,
        View<C>,
        View<D>,
        View<E>,
        View<F>,
        View<G>,
        View<H>,
        View<I>,
        View<J>,
    ),
    k: View<K>,
) {}
```

You can split it any way you want, often components are more closely related to a few others and you end up with multiple small tuples.

Another way is to use custom views.

```rust
#[derive(Borrow, BorrowInfo)]
struct SysViews<'v> {
    a: View<'v, A>,
    b: View<'v, B>,
    c: View<'v, C>,
    d: View<'v, D>,
    e: View<'v, E>,
    f: View<'v, F>,
    g: View<'v, G>,
    h: View<'v, H>,
    i: View<'v, I>,
    j: View<'v, J>,
    k: View<'v, K>,
}

fn sys(views: SysViews) {}

// or destructure

fn sys(
    SysViews {
        a,
        b,
        c,
        d,
        e,
        f,
        g,
        h,
        i,
        j,
        k,
    }: SysViews,
) {}
```

Of course you don't have to put all of them in the custom view. And it doesn't have to be tied to one system specifically. If a few views are (almost) always borrowed together, it could be interesting to create a custom view for them.

----

For `bulk_add_entity`, there isn't any tricks. What you can do is split the call in two.
You can simply the view borrowing with one the method above. And/or use `World::borrow` instead of `World::run`.

```rust
fn create_boids(world: &mut World, team: Team, n: u64, center: Vec2, radius: f32) {
    let mut next_id = BOID_ID
        .fetch_update(Ordering::SeqCst, Ordering::SeqCst, |v| Some(v + n))
        .unwrap_ctx("Updating BOID_ID");

    world.run(
        |mut entities: EntitiesViewMut,
         (boids, pos, boid_dir, boid_energy): (
            ViewMut<Boid>,
            ViewMut<Pos>,
            ViewMut<BoidDir>,
            ViewMut<BoidEnergy>,
        ),
         mut near_boids: ViewMut<NearBoids>,
         mut pray_for_spirits: ViewMut<PrayForSpirit>,
         mut avoidance_force: ViewMut<AvoidanceForce>,
         mut cohesion_force: ViewMut<CohesionForce>,
         mut alignment_force: ViewMut<AlignmentForce>,
         mut seperation_force: ViewMut<SeparationForce>,
         mut drawable: ViewMut<Drawable>| {
            let iter = entities.bulk_add_entity(
                (boids, pos, boid_dir, boid_energy),
                (0..n).map(|_| {
                    let id = next_id + i;
                    let pos = (center + radial_random_pos(radius)).round();

                    let dir = Vec2::new(
                        random::gen::<f32>() * 2.0 - 1.0,
                        random::gen::<f32>() * 2.0 - 1.0,
                    )
                    .normalize_or_zero();

                    let boid = Boid {
                        id,
                        team,
                        ..Default::default()
                    };

                    (
                        boid,
                        Pos(pos),
                        BoidDir(dir),
                        BoidEnergy(random::range(0.0..boid.max_energy)),
                    )
                }),
            );

            near_boids.reserve(n);
            pray_for_spirits.reserve(n);
            avoidance_force.reserve(n);
            cohesion_force.reserve(n);
            alignment_force.reserve(n);
            seperation_force.reserve(n);
            drawable.reserve(n);

            for entity in iter {
                near_boids.add_component_unchecked(entity, NearBoids::default());
                pray_for_spirits.add_component_unchecked(entity, PrayForSpirit::default());
                avoidance_force.add_component_unchecked(entity, AvoidanceForce(Vec2::ZERO));
                cohesion_force.add_component_unchecked(entity, CohesionForce(Vec2::ZERO));
                alignment_force.add_component_unchecked(entity, AlignmentForce(Vec2::ZERO));
                seperation_force.add_component_unchecked(entity, SeparationForce(Vec2::ZERO));
                drawable.add_component_unchecked(entity, Drawable::Soul { anim_delta: 0.0 });
            }
        },
    );
}
```

**Comment by Nazariglez** [comment]
The query solution is great, the adding components part doesn't look very ergonomic but thats fine. I can keep working with this. Thanks!

As a side note, I tried to create a derive macro to impl the required traits using the field list, to do something like this:

```rust 
#[derive(ComponentList)]
struct BoidComponentList {
    boid: Boid,
    pos: Pos,
    dir: BoidDir,
    energy: BoidEnergy,
    near_boids: NearBoids,
    pray: PrayForSpirit,
    avoidance_force: AvoidanceForce,
    cohesion_force: CohesionForce,
    alignment_force: AlignmentForce,
    separation_force: SeparationForce,
    drawable: Drawable,
}

fn create_boids(world: &mut World, team: Team, n: u64, center: Vec2, radius: f32) {
    let mut next_id = BOID_ID
        .fetch_update(Ordering::SeqCst, Ordering::SeqCst, |v| Some(v + n))
        .unwrap_ctx("Updating BOID_ID");

    world.bulk_add_entity((0..n).map(|i| {
        let id = next_id + i;
        let pos = (center + radial_random_pos(radius)).round();

        let dir = Vec2::new(
            random::gen::<f32>() * 2.0 - 1.0,
            random::gen::<f32>() * 2.0 - 1.0,
        )
        .normalize_or_zero();

        let boid = Boid {
            id,
            team,
            ..Default::default()
        };

        BoidComponentList {
            boid,
            pos: Pos(pos),
            dir: BoidDir(dir),
            energy: BoidEnergy(random::range(0.0..boid.max_energy)),
            near_boids: NearBoids::default(),
            pray: PrayForSpirit::default(),
            avoidance_force: AvoidanceForce::default(),
            cohesion_force: CohesionForce::default(),
            alignment_force: AlignmentForce::default(),
            separation_force: SeparationForce::default(),
            drawable: Drawable::Soul { anim_delta: 0.0 },
        }
    }));
```

However, the methods/traits needed to do this in a similar way to the `bulk_insert` macro are private, and I'm not sure if it makes sense. Do you have any thoughts on this?

**Comment by leudz** [comment]
I checked the same thing and noticed the private fields in the returned iterator.

I think it'll make a good addition to sidestep the lack of variadics in this case.

**Comment by Nazariglez** [comment]
I'm not currently blocked by this thanks to your solutions, but I can try creating a PR for it if you don't have anything in progress.

**Comment by leudz** [comment]
I haven't started working on it.
A PR would be greatly appreciated. 
I can help if you have any question.

**Comment by Nazariglez** [comment]
Got it, I'll take a look then

**Comment by Nazariglez** [comment]
This could be related so I am using same issue, I cannot iterate over elements when using more than 10 elements, the system works fine with nested tuples to get the storages but when trying to iterate I tried with:

```rust
        (
            &boids,
            &mut dir,
            &mut pos,
            &energy,
            &avoid,
            &cohesion,
            &align,
            &separation,
            &chase,
            !&resting,
            !&totem_slots,
        ),
--
        (
            &boids,
            &mut dir,
            &mut pos,
            &energy,
            (&avoid, &cohesion, &align, &separation, &chase),
            !&resting,
            !&totem_slots,
        ),
--
#[derive(Borrow, BorrowInfo, IntoIter)]
pub struct MyMoveSystemView<'v> {
    #[shipyard(item_field_name = "boid")]
    boids: View<'v, Boid>,
    #[shipyard(item_field_name = "dir")]
    dirs: ViewMut<'v, BoidDir>,
    #[shipyard(item_field_name = "pos")]
    positions: ViewMut<'v, Pos>,
    #[shipyard(item_field_name = "energy")]
    energies: View<'v, BoidEnergy>,
    #[shipyard(item_field_name = "avoid")]
    avoidance_force: View<'v, AvoidanceForce>,
    #[shipyard(item_field_name = "cohesion")]
    cohesion_force: View<'v, CohesionForce>,
    #[shipyard(item_field_name = "alignment")]
    alignment_force: View<'v, AlignmentForce>,
    #[shipyard(item_field_name = "separation")]
    separation_force: View<'v, SeparationForce>,
    #[shipyard(item_field_name = "chase")]
    chase_force: View<'v, ChaseForce>,
}

pub fn move_system(
    mut sys_view: MyMoveSystemView,
    resting: View<BoidResting>,
    gd: UniqueView<GlobalData>,
) {
    let dt = gd.fixed_delta;

    (&mut sys_view, !&resting).iter().for_each(|_| {});
}
```

With same results:

```
rustc: the method `iter` exists for tuple `*`, but its trait bounds were not satisfied
```

Any idea? Thanks

**Comment by leudz** [comment]
The 10 storages limit is not an inherent limitation of the proc macro, only one of the current implementation.
In theory, the proc macro should replicate this macro rules (and a little more):
https://github.com/leudz/shipyard/blob/82e6e305e8ca3dc41ca69603d57d97d8378a3143/src/iter/into_iter.rs#L177-L343

Instead, it relies on iterator types that `impl_into_iter` generates, making it inherit the 10 storages limit.

The reason for this is maintain cost.
The `IntoIter` proc macro is not super old. At that time, it was already clear that the whole iter module had to be reworked.
One of the goal of the rework is to allow custom views to become iterable (a lot) simpler and compose with other views.
The lack of composition is why `(&mut sys_view, !&resting).iter().for_each(|_| {});` doesn't work, custom views don't implement the right trait to allow this.
But it requires a better design for `IntoAbstract` and `AbstractMut`, that is simpler without loosing on the performance.
It's one of the goal for the next version, I don't know when that will be yet.

----

As you already found, one workaround is to merge components. Usually if components are (almost) always borrowed together, it's a good sign they should be merged.
In the same vein, splitting the system/loop is some times possible.

One thing to know about `SparseSet` iteration is that it's basically this ([there is a chapter that explains more this](https://leudz.github.io/shipyard/guide/master/going-deeper/sparse-set.html)):
```rust
for entity_id in entities {
    let Some(comp1) = storages1.get(entity_id) else {
        continue;
    }
    let Some(comp2) = storages1.get(entity_id) else {
        continue;
    }
    let Some(comp3) = storages1.get(entity_id) else {
        continue;
    }
    // ...
}
```

Since we don't want to visit all entities, instead one of the storage is picked as the "driver".

```rust
for entity_id in driver {
    let Some(comp1) = storages1.get(entity_id) else {
        continue;
    }
    let Some(comp2) = storages1.get(entity_id) else {
        continue;
    }
    let Some(comp3) = storages1.get(entity_id) else {
        continue;
    }
    // ...
}
```

The driver is the storage with the least components, and it cannot be a `!view`.
Tracking storage are also not picked as much because they're slower to iterate.

You don't seem to do anything with tracking here, so we can ignore that.
You can iterate manually any number of storages by doing this:
```rust
fn move_system(
    mut sys_view: MyMoveSystemView,
    resting: View<BoidResting>,
    gd: UniqueView<GlobalData>,
) {
    let dt = gd.fixed_delta;

    for (eid, boids) in sys_view.boids.iter().with_id() {
        if resting.get(eid).is_ok() {
            continue;
        }

        let Ok((
            dir,
            pos,
            energy,
            avoidance_force,
            cohesion_force,
            alignment_force,
            separation_force,
            chase_force,
        )) = (
            &mut sys_view.dirs,
            &mut sys_view.positions,
            &sys_view.energies,
            &sys_view.avoidance_force,
            &sys_view.cohesion_force,
            &sys_view.alignment_force,
            &sys_view.separation_force,
            &sys_view.chase_force,
        )
            .get(eid)
        else {
            continue;
        };
    }
}
```

I'm assuming all storages have the same number of components, if not then it's faster to pick the one with the least components as driver.
I put as many storages in the `get` as possible but you can separate the calls instead.
It's definitely not as clean as the macro, but it will work with any number of storage with the current version.

**Comment by Nazariglez** [comment]
Thanks @leudz. Let's close this for now, I am still working on the game with the goal to publish it in the next months, so I'll probably come back to the bundle thing with a PR eventually, but not now. (if you're interested, I share the progress here https://x.com/Nazariglez/status/1847229678811033867). 

I have two more questions that are not really important, but I would love to see what you think. Is it possible to use `with_id` with `par_iter` somehow? And, how about using `Not` in a query `!&storage` without borrow the storage to be used inside of the system (ie: adding a component)?

**Comment by leudz** [comment]
Looks interesting, definitely send a link when it's available!

For parallel `with_id` it's not currently possible. It's not impossible but a few things would have to change. The whole `LastId` trait is a very bad design so likely it would be another set of trait methods on `AbstractMut`.

For `Not` it's not simple. Shipyard has to borrow the storage to check inside if the component exist or not. ECS based on archetypes, like Bevy for example, don't have to because they literally store the component in another storage in that case.
But in theory a lot rust collections are too restrictive when it comes to borrow. It's ok to iterate and push to a `Vec` as long as it doesn't reallocate. The same could happen here.
It would be a kind of `split_at(_mut)` with the right half being uninitialized. Then you could fill the remaining `capacity - length` elements. Apparently there is a crate for it [split_at_extra_cap](https://docs.rs/uninit/latest/uninit/extension_traits/trait.VecCapacity.html#method.split_at_extra_cap).

Without that, you could take the storage out if the iterator, do a manual `if storage.contains(entity) { continue; }` and then you can push. It's a dirty workaround, as usual 😅
